reactive/kotlinx-coroutines-reactive/src/Publish.kt

```diff
@@ -84,6 +81,39 @@ public class PublisherCoroutine<in T>(
     override fun invokeOnClose(handler: (Throwable?) -> Unit): Nothing =
         throw UnsupportedOperationException("PublisherCoroutine doesn't support invokeOnClose")
 
+    // Mutex is locked when either nRequested == 0 or while subscriber.onXXX is being invoked
+    private val mutex: Mutex = PublisherCoroutineMutex()
+
+    /**
+     * To send an element, [mutex] should be locked first, after which [doLockedNext]
+     * is called -- it releases the mutex at the end. To support the `select` operation
+     * and implement [onSend], we use a modified version of [Mutex.onLock], which
+     * invokes [doLockedNext] at the end and returns this [PublisherCoroutine] as result
+     * (see [PublisherCoroutineMutex.onLockProcessResult]).
+     *
+     * We use this dirty hack as we need to wait for a lock, and, therefore, the clause
+     * object should be a [Mutex] instance. Thus, the [doLockedNext] call must be a part
+     * of [PublisherCoroutineMutex.onLockProcessResult] as well. A possible alternative
+     * would be to inherit [PublisherCoroutine] from [Mutex], but it already extends [AbstractCoroutine].
+     */
+    @Suppress("INVISIBLE_MEMBER", "INVISIBLE_REFERENCE", "CANNOT_OVERRIDE_INVISIBLE_MEMBER")
```

Comment:
I finally see what's going on here.

I suggest reverting this and fallback to previous behaviour: `onLock` is deprecated **for removal**. It means that eventually it at least will be hidden and no amount of suppresses will help us, forcing us to fallback to the previous hack with a coroutine anyways.

The earlier we start using such code, the better (so any potential regression will be found, the current workaround won't be forgotten and so on). Bonus points -- we won't have updating this code when raising deprecation levels.

Reply:
Why do you want to completely remove `onLock` when its support for binary compatibility is very cheap with this new `select` implementation?

Reply:
It's always cheaper to remove things when possible, because eventually `select` implementation (or any other underlying primitive) will be changed or tweaked, but as long as `onLock` is deprecated with ERROR+, we won't be able to test it properly.

Even if we'll decide not to remove it, it definitely will be promoted to `HIDDEN` deprecation and we won't be able to use it here, so the fallback is required

Reply:
OK, it is better then to refactor the code another way, using the registration function directly.

Reply:
The code overrides `onLockRegFunction` and `onLockProcessResult` functions, which are internal. After promoting the deprecation level for `onLock`, its usage can be replaced with an internal `onLockInternal`. 

We can keep the previous workaround, but it may introduce unpredictable behavior due to starting a new coroutine (even in the UNDISPATCHED mode), similar to the old `select` API implementation via the new one.

Reply:
I did not know that this code is marginal. In this case, I agree with your point. I did a similar trick, please check the code.